File: day2/py/puzzle2.py
#!/usr/bin/env python

ABC = { 'A': 1, 'B': 2, 'C': 3}
XYZ = { 'X': 0, 'Y': 3, 'Z': 6}

lose_draw_win = {
    'X': { 'A': 'C', 'B': 'A', 'C': 'B' },
    'Y': { 'A': 'A', 'B': 'B', 'C': 'C' },
    'Z': { 'A': 'B', 'B': 'C', 'C': 'A' }
}

def solve():
    score = 0

    with open('../input.txt', 'r') as f:
        for line in f:
            opponent = line[0]
            outcome = line[2]

            score += ABC[lose_draw_win[outcome][opponent]] + XYZ[outcome]        

    return score

if __name__ == "__main__":
    solution = solve()
    print(solution, )

File: day2/cxx/puzzle2.cxx
#include <algorithm>
#include <fstream>
#include <iostream>
#include <unordered_map>
#include <string>

#include "lines.hpp"

using namespace std;

static unordered_map<string, char> lose_draw_win {
    {"A X", 'C'}, {"A Y", 'A'}, {"A Z", 'B'},
    {"B X", 'A'}, {"B Y", 'B'}, {"B Z", 'C'},
    {"C X", 'B'}, {"C Y", 'C'}, {"C Z", 'A'},
};

const auto solve()
{
    ifstream f{"../input.txt"};

    auto score = 0u;

    for (auto & line: lines(f))
    {
        auto choice = (lose_draw_win[line.substr(0,3)] - 'A' + 1);
        auto outcome = ((line[2] - 'X') * 3); 
        score += choice + outcome;
    }

    return score;
}

int main()
{
    auto solution = solve();
    cout << solution << endl;
}